### PR TITLE
Error on numpy masked array inputs.

### LINKED ISSUE #13795: JAX silently discards masks in numpy masked arrays
Reported in https://github.com/data-apis/array-api/issues/582
```python
>>> jnp.sin(np.ma.array([1, 2, 3], mask=[True, False, True]))
Array([0.84147096, 0.9092974 , 0.14112   ], dtype=float32)
```

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #13795 - there are some additional changes required for the C++ `device_put` path; I'll address those in a separate change.

### CODE REVIEW

File: jax/_src/abstract_arrays.py [hunk 1]
```diff
@@ -51,6 +51,13 @@ def canonical_concrete_aval(val, weak_type=None):
   return ConcreteArray(dtypes.canonicalize_dtype(np.result_type(val)), val,
                        weak_type=weak_type)
 
+def masked_array_error(*args, **kwargs):
+  raise ValueError("numpy masked arrays are not supported as direct inputs to JAX functions. "
```
Comment (froystig) [inline jax/_src/abstract_arrays.py:55]:
```suggestion
  raise ValueError("numpy masked arrays are not supported as direct inputs to `jax.numpy` functions. "
```
?

File: jax/_src/dispatch.py [hunk 2]
```diff
@@ -1275,6 +1275,10 @@ def device_put(x, device: Optional[Device] = None) -> Tuple[Any, ...]:
 # TODO(phawkins): update users.
 xla.device_put = device_put
 
+def _device_put_masked_array(x, device: Optional[Device]):
+  raise ValueError("numpy masked arrays are not supported as direct inputs to JAX functions. "
```
Comment (froystig) [inline jax/_src/dispatch.py:1279]:
```suggestion
  raise ValueError("numpy masked arrays are not supported as direct inputs to `jax.numpy` functions. "
```
?

Comment (froystig) [reply jax/_src/dispatch.py:1279]: etc. (several occurrences below)

**Review by froystig — APPROVED** [review]

File: jax/_src/abstract_arrays.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/abstract_arrays.py:55]: I think the more general form is correct: it's not just `jax.numpy` functions, but *any* jax function (e.g. `jax.lax`, anything transformed with `jax.jit` or `jax.vmap`, etc.)
